php-checks/src/main/java/org/sonar/php/checks/UseOfUninitializedVariableCheck.java

```diff
@@ -0,0 +1,279 @@
+    checkFunction(tree);
+    super.visitFunctionDeclaration(tree);
+  }
+
+  @Override
+  public void visitFunctionExpression(FunctionExpressionTree tree) {
+    checkFunction(tree);
+    super.visitFunctionExpression(tree);
+  }
+
+  @Override
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    checkFunction(tree);
+    super.visitMethodDeclaration(tree);
+  }
+
+  private void checkFunction(FunctionTree function) {
+    VariableVisitor visitor = new VariableVisitor(function);
+    function.accept(visitor);
+    visitor.uninitializedStream()
+      .forEach(variable -> context().newIssue(this, variable, "Review the data-flow - use of uninitialized value."));
+  }
+
+  private static class VariableVisitor extends PHPVisitorCheck {
+
+    final FunctionTree currentFunction;
+
+    boolean trustVariables = true;
+
+    Map<String, VariableIdentifierTree> firstVariableReadAccess = new HashMap<>();
+
+    Set<String> initializedVariables = new HashSet<>();
+
+    VariableVisitor(FunctionTree currentFunction) {
+      this.currentFunction = currentFunction;
+    }
+
+    Stream<VariableIdentifierTree> uninitializedStream() {
+      if (!trustVariables) {
+        return Stream.empty();
+      }
+      return firstVariableReadAccess.entrySet().stream()
+        .filter(entry -> !initializedVariables.contains(entry.getKey()))
+        .map(Map.Entry::getValue);
+    }
+
+    @Override
+    public void visitVariableIdentifier(VariableIdentifierTree tree) {
+      if (isMemberAccess(tree)) {
+        return;
+      }
+      String lowerCaseName = tree.text().toLowerCase(Locale.ROOT);
+      if (!PREDEFINED_VARIABLES.contains(lowerCaseName)) {
+        if (isReadAccess(tree)) {
+          if (!firstVariableReadAccess.containsKey(lowerCaseName)) {
+            firstVariableReadAccess.put(lowerCaseName, tree);
+          }
+        } else {
+          initializedVariables.add(lowerCaseName);
+        }
+      }
+    }
+
+    @Override
+    public void visitFunctionCall(FunctionCallTree functionCall) {
+      if (FUNCTION_CHANGING_CURRENT_SCOPE.contains(lowerCaseFunctionName(functionCall))) {
+        trustVariables = false;
+      }
+      super.visitFunctionCall(functionCall);
+    }
+
+    @Override
+    public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
+      if (tree == currentFunction) {
+        super.visitFunctionDeclaration(tree);
+      }
+      // else skip nested
+    }
+
+    @Override
+    public void visitFunctionExpression(FunctionExpressionTree tree) {
+      LexicalVariablesTree lexicalVars = tree.lexicalVars();
+      if (tree == currentFunction) {
+        if (lexicalVars != null) {
+          lexicalVars.variables().stream()
+            .map(VariableTree::variableExpression)
+            .filter(variable -> variable.is(Kind.VARIABLE_IDENTIFIER))
+            .map(variable -> ((VariableIdentifierTree) variable).text().toLowerCase(Locale.ROOT))
```

Comment:
I think we shouldn't transform the name of the variable. Variable names are case-sensitive.